File: android-client/app/src/main/kotlin/com/lowremote/session/RemoteSession.kt
package com.lowremote.session

import android.content.Context
import android.media.AudioFormat
import android.media.AudioRecord
import android.media.AudioTrack
import android.media.MediaRecorder
import android.util.Log
import android.view.Surface
import com.lowremote.audio.AudioPlayer
import com.lowremote.codec.FrameAssembler
import com.lowremote.codec.H265Decoder
import com.lowremote.model.ControlEvent
import com.lowremote.model.Packet
import com.lowremote.model.RemoteDevice
import com.lowremote.network.TcpClient
import com.lowremote.network.UdpReceiver
import com.lowremote.network.UdpSender
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

class RemoteSession {

    companion object {
        private const val TAG = "RemoteSession"
        private const val CLIENT_UDP_PORT = 0
        private const val HEARTBEAT_INTERVAL_MS = 30_000L
    }

    enum class State { Idle, Connecting, Connected, Disconnected }

    data class ScreenInfo(val index: Int, val name: String, val width: Int, val height: Int)

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Default)

    private val tcp      = TcpClient(scope)
    private val receiver = UdpReceiver(port = CLIENT_UDP_PORT)
    private val sender   = UdpSender()
    private val assembler = FrameAssembler()

    // ── Mac-system-audio player (type 0x04) ────────────────────────────────────
    private val macAudioPlayer = AudioPlayer()

    private var decoder: H265Decoder? = null
    @Volatile private var surface: Surface? = null
    private var heartbeatJob: Job? = null
    private val decoderLock = Any()
    private var device: RemoteDevice? = null
    private var eventFrameId: Int = 0

    // ── Audio ──────────────────────────────────────────────────────────────────
    private var audioRecord: AudioRecord? = null
    private var audioJob: Job? = null

    // ── State flows ────────────────────────────────────────────────────────────
    private val _state             = MutableStateFlow(State.Idle)
    val state: StateFlow<State>    = _state

    private val _remoteResolution  = MutableStateFlow<Pair<Int, Int>?>(null)
    val remoteResolution: StateFlow<Pair<Int, Int>?> = _remoteResolution

    private val _fps               = MutableStateFlow(60)
    val fps: StateFlow<Int>        = _fps

    private val _screens           = MutableStateFlow<List<ScreenInfo>>(emptyList())
    val screens: StateFlow<List<ScreenInfo>> = _screens

    private val _currentScreen     = MutableStateFlow(0)
    val currentScreen: StateFlow<Int> = _currentScreen

    private val _audioEnabled      = MutableStateFlow(false)
    val audioEnabled: StateFlow<Boolean> = _audioEnabled

    // ── Init ───────────────────────────────────────────────────────────────────
    init {
        assembler.onFrameReady = { bytes, isKeyframe -> decoder?.feed(bytes, isKeyframe) }
        receiver.onPacket = { parsed, payload, _ ->
            when (parsed.type) {
                Packet.TYPE_VIDEO        -> assembler.onPacket(parsed, payload)
                Packet.TYPE_SYSTEM_AUDIO -> macAudioPlayer.write(payload)
            }
        }
        tcp.onLine       = { line -> handleTcpLine(line) }
        tcp.onDisconnected = { scope.launch { teardown() } }
    }

    // ── Surface ────────────────────────────────────────────────────────────────
    fun setSurface(s: Surface?) {
        synchronized(decoderLock) {
            surface = s
            if (s == null) { decoder?.stop(); decoder = null; return }
            if (_state.value == State.Connected && decoder == null) startDecoderIfReadyLocked()
        }
    }

    // ── Connect ────────────────────────────────────────────────────────────────
    fun connect(device: RemoteDevice, fps: Int) {
        if (_state.value != State.Idle && _state.value != State.Disconnected) return
        _state.value = State.Connecting
        this.device  = device
        _fps.value   = fps

        scope.launch {
            receiver.start()
            val sock = receiver.sharedSocket() ?: run { teardown(); return@launch }
            sender.attach(sock, device.host, device.udpPort)
            sender.sendEvent("HELLO", nextEventFrameId())

            if (!tcp.connect(device.host, device.tcpPort)) { teardown(); return@launch }

            tcp.send("FPS:$fps")
            _state.value = State.Connected
            macAudioPlayer.start()
            startHeartbeat()
        }
    }

    fun changeFps(fps: Int) {
        _fps.value = fps
        if (_state.value == State.Connected) {
            synchronized(decoderLock) { decoder?.flush() }
            assembler.reset()
            tcp.send("FPS:$fps")
        }
    }

    fun switchScreen(index: Int) {
        if (_state.value != State.Connected) return
        _currentScreen.value = index
        // Tear down old decoder — resolution may change
        synchronized(decoderLock) { decoder?.stop(); decoder = null }
        assembler.reset()
        tcp.send("SCREEN:$index")
    }

    fun sendEvent(event: ControlEvent) {
        if (_state.value != State.Connected) return
        val serialized = event.serialize()
        val frameId    = nextEventFrameId()
        scope.launch(Dispatchers.IO) {
            try { sender.sendEvent(serialized, frameId) }
            catch (e: Exception) { Log.w(TAG, "sendEvent failed: ${e.message}") }
        }
    }

    // ── Audio ──────────────────────────────────────────────────────────────────
    /**
     * Toggle microphone capture and streaming to Mac.
     *
     * Protocol:
     *   1. Android sends TCP "AUDIO_ON"  → Mac starts AVAudioEngine playback
     *   2. Android streams raw PCM chunks over UDP (type=0x03)
     *      Format: 16 000 Hz, mono, 16-bit signed little-endian, ~20 ms chunks
     *   3. Mac plays PCM through its default output device.
     *      All apps that read from the default *input* will NOT hear this directly;
     *      the audio comes out of the speakers/headphones.
     *      To route it as a microphone input you need a loopback virtual device
     *      (e.g. BlackHole 2ch) set as both output + input in Audio MIDI Setup.
     *      For **dictation / Siri / speech recognition** on macOS, set the Mac's
     *      System Settings → Sound → Input to the same virtual device.
     *
     * Note: RECORD_AUDIO permission must be granted before calling this.
     */
    fun setAudioEnabled(enabled: Boolean) {
        if (enabled == _audioEnabled.value) return
        _audioEnabled.value = enabled
        if (enabled && _state.value == State.Connected) {
            tcp.send("AUDIO_ON")   // tell Mac to start AVAudioEngine
            startAudioCapture()
        } else {
            stopAudioCapture()
            if (_state.value == State.Connected) tcp.send("AUDIO_OFF")
        }
    }

    private fun startAudioCapture() {
        stopAudioCapture()
        // Must match Mac AudioReceiver exactly:
        //   sampleRate = 16 000 Hz
        //   channels   = mono (1)
        //   encoding   = PCM 16-bit signed little-endian
        val sampleRate = 16_000
        val channelConfig = AudioFormat.CHANNEL_IN_MONO
        val encoding = AudioFormat.ENCODING_PCM_16BIT

        // Use ~20 ms chunks: 16000 samples/s * 0.02 s * 2 bytes = 640 bytes.
        // getMinBufferSize is a lower bound; we use 4× to avoid underruns.
        val minBuf = AudioRecord.getMinBufferSize(sampleRate, channelConfig, encoding)
        if (minBuf == AudioRecord.ERROR_BAD_VALUE || minBuf <= 0) {
            Log.w(TAG, "AudioRecord.getMinBufferSize failed"); return
        }
        val bufSize = maxOf(minBuf * 4, 1280) // at least 2 × 20ms frames

        val rec = try {
            AudioRecord(
                MediaRecorder.AudioSource.MIC,
                sampleRate, channelConfig, encoding, bufSize
            )
        } catch (e: Exception) { Log.w(TAG, "AudioRecord ctor: ${e.message}"); return }

        if (rec.state != AudioRecord.STATE_INITIALIZED) {
            Log.w(TAG, "AudioRecord not initialized"); rec.release(); return
        }

        rec.startRecording()
        audioRecord = rec
        Log.d(TAG, "audio capture started: ${sampleRate} Hz mono 16-bit, buf=$bufSize")

        audioJob = scope.launch(Dispatchers.IO) {
            // Each read delivers one chunk; we send it immediately.
            // 20-ms read size keeps latency low while amortising UDP overhead.
            val chunkBytes = (sampleRate / 50) * 2   // 20 ms × 2 bytes/sample = 640
            val buf = ByteArray(chunkBytes)
            while (isActive && _audioEnabled.value) {
                val read = rec.read(buf, 0, chunkBytes)
                if (read > 0) {
                    val packet = Packet.encodeAudio(nextEventFrameId(), buf.copyOf(read))
                    sender.send(packet)
                }
            }
        }
    }

    private fun stopAudioCapture() {
        audioJob?.cancel(); audioJob = null
        try { audioRecord?.stop() } catch (_: Exception) {}
        audioRecord?.release(); audioRecord = null
    }

    // ── Disconnect / Release ───────────────────────────────────────────────────
    fun disconnect() {
        if (_state.value == State.Idle) return
        if (tcp.isConnected) tcp.send("DISCONNECT")
        scope.launch { teardown() }
    }

    fun release() {
        scope.launch { teardown() }
        scope.cancel()
    }

    // ── TCP line handler ───────────────────────────────────────────────────────
    private fun handleTcpLine(line: String) {
        val trimmed = line.trim()
        when {
            trimmed.startsWith("RESOLUTION:") -> {
                val p = trimmed.removePrefix("RESOLUTION:").split(",")
                val w = p.getOrNull(0)?.toIntOrNull()
                val h = p.getOrNull(1)?.toIntOrNull()
                if (w != null && h != null) {
                    _remoteResolution.value = w to h
                    startDecoderIfReady()
                }
            }
            trimmed.startsWith("SCREENS:") -> {
                // Format: SCREENS:0:主屏幕:2560x1600,1:屏幕2:1920x1080
                val list = trimmed.removePrefix("SCREENS:").split(",").mapNotNull { entry ->
                    val parts = entry.split(":")
                    if (parts.size < 3) return@mapNotNull null
                    val idx  = parts[0].toIntOrNull() ?: return@mapNotNull null
                    val name = parts[1]
                    val dim  = parts[2].split("x")
                    val w2   = dim.getOrNull(0)?.toIntOrNull() ?: 0
                    val h2   = dim.getOrNull(1)?.toIntOrNull() ?: 0
                    ScreenInfo(idx, name, w2, h2)
                }
                if (list.isNotEmpty()) _screens.value = list
            }
            trimmed == "OK" -> startDecoderIfReady()
            trimmed == "PONG" -> { /* heartbeat */ }
        }
    }

    // ── Decoder ────────────────────────────────────────────────────────────────
    private fun startDecoderIfReady() {
        synchronized(decoderLock) { startDecoderIfReadyLocked() }
    }
    private fun startDecoderIfReadyLocked() {
        val s        = surface ?: return
        if (decoder != null) return
        val (w, h)   = _remoteResolution.value ?: (1920 to 1080)
        val dec      = H265Decoder(s, _fps.value)
        dec.start(w, h)
        decoder      = dec
        Log.d(TAG, "decoder: ${w}x${h} @ ${_fps.value}fps")
    }

    // ── Heartbeat ──────────────────────────────────────────────────────────────
    private fun startHeartbeat() {
        heartbeatJob?.cancel()
        heartbeatJob = scope.launch {
            while (isActive) {
                delay(HEARTBEAT_INTERVAL_MS)
                if (tcp.isConnected) tcp.send("PING") else break
            }
        }
    }

    // ── Teardown ───────────────────────────────────────────────────────────────
    private suspend fun teardown() {
        heartbeatJob?.cancel(); heartbeatJob = null
        if (_audioEnabled.value) tcp.send("AUDIO_OFF")
        stopAudioCapture()
        _audioEnabled.value = false
        macAudioPlayer.stop()
        tcp.disconnect()
        receiver.stop()
        synchronized(decoderLock) { decoder?.stop(); decoder = null }
        assembler.reset()
        _state.value = State.Disconnected
        device = null
    }

    private fun nextEventFrameId(): Int {
        eventFrameId = (eventFrameId + 1) and 0x7FFFFFFF
        return eventFrameId
    }
}

File: ios-client/LowRemote/Session/RemoteSession.swift
import Foundation
import UIKit
import SwiftUI

/// 远程会话生命周期管理，完整对齐 Android RemoteSession.kt
///
/// 负责：TCP 控制通道、UDP 视频接收、帧重组、解码、音频双向、文件传输、心跳
@Observable
final class RemoteSession {

    // MARK: - State

    enum State: Equatable {
        case idle
        case connecting
        case connected
        case disconnected
    }

    private(set) var state:            State          = .idle
    private(set) var remoteResolution: CGSize?        = nil
    private(set) var fps:              Int            = 60
    private(set) var screens:          [ScreenInfo]   = []
    private(set) var currentScreen:    Int            = 0
    private(set) var audioEnabled:     Bool           = false
    private(set) var fileTransferProgress: Double?    = nil   // 0.0-1.0，nil=空闲

    struct ScreenInfo: Identifiable {
        let id:     Int
        let name:   String
        let width:  Int
        let height: Int
    }

    // MARK: - Components

    private let tcp       = TcpClient()
    private let receiver  = UdpReceiver(port: 0)      // 0 = 系统分配端口
    private let sender    = UdpSender()
    private let assembler = FrameAssembler()
    private var decoder:  H265Decoder?

    private let macAudioPlayer = AudioPlayer()
    private let audioCapture   = AudioCapture()

    // 视频渲染视图（由 VideoSurface SwiftUI wrapper 注入）
    private weak var videoView: VideoSurfaceView?

    private var device: RemoteDevice?
    private var eventFrameId: UInt32 = 0

    // 心跳
    private var heartbeatTask: Task<Void, Never>?

    // 文件传输（原子标志位）
    private var fileReadySignal: Bool = false
    private let fileReadyLock = NSLock()

    // 后台任务
    private var bgTask: UIBackgroundTaskIdentifier = .invalid

    // 解码器锁
    private let decoderLock = NSLock()

    // MARK: - Init

    init() {
        assembler.onFrameReady = { [weak self] data, isKeyframe in
            self?.decoder?.feed(data: data, isKeyframe: isKeyframe)
        }
        receiver.onPacket = { [weak self] parsed, payload, _ in
            switch parsed.type {
            case Packet.typeVideo:
                self?.assembler.onPacket(parsed, payload)
            case Packet.typeSystemAudio:
                self?.macAudioPlayer.write(payload)
            default:
                break
            }
        }
        tcp.onLine = { [weak self] line in
            self?.handleTcpLine(line)
        }
        tcp.onDisconnected = { [weak self] in
            Task { await self?.teardown() }
        }
        audioCapture.onAudioChunk = { [weak self] pcm in
            guard let self = self else { return }
            self.sender.sendAudio(pcm, frameId: self.nextFrameId())
        }
    }

    // MARK: - Video Surface

    func setVideoView(_ view: VideoSurfaceView?) {
        decoderLock.lock()
        defer { decoderLock.unlock() }
        videoView = view
        guard let v = view else { return }
        // 更新已运行解码器的渲染目标
        if let dec = decoder {
            dec.displayLayer = v.displayLayer
            return
        }
        // 解码器尚未启动：若已连接则立即启动（补偿 RESOLUTION/OK 到达时 view 还未注入的竞态）
        if state == .connected {
            startDecoderLocked(view: v)
        }
    }

    // MARK: - Connect

    func connect(device: RemoteDevice, fps: Int = 60) {
        guard state == .idle || state == .disconnected else { return }
        state      = .connecting
        self.device = device
        self.fps    = fps

        Task {
            await performConnect(device: device, fps: fps)
        }
    }

    private func performConnect(device: RemoteDevice, fps: Int) async {
        beginBackgroundTask()

        receiver.start()
        sender.attach(fd: receiver.rawFd, host: device.host, port: device.udpPort)

        // 先发 HELLO，让 Mac 记录客户端 UDP 端点（clientEndpoint）
        // 必须等 HELLO 实际送出后再建立 TCP 连接并发 FPS 命令
        // 否则 Mac 收到 FPS 时 clientEndpoint=nil，startStreaming 产生的帧全部丢弃
        sender.sendEvent("HELLO", frameId: nextFrameId())
        // 等待底层 sendto 完成：sender 内部用 async queue，给 50ms 让 HELLO 包离开发送队列
        try? await Task.sleep(nanoseconds: 50_000_000)

        let ok = await tcp.connect(host: device.host, port: device.tcpPort)
        guard ok else {
            await teardown(); return
        }

        tcp.send("FPS:\(fps)")
        macAudioPlayer.start()

        await MainActor.run {
            state = .connected
        }
        startHeartbeat()
    }

    // MARK: - Controls

    func changeFps(_ newFps: Int) {
        fps = newFps
        guard state == .connected else { return }
        // flush 在后台执行避免主线程阻塞
        decoderLock.lock()
        let dec = decoder
        decoderLock.unlock()
        if let dec = dec {
            DispatchQueue.global(qos: .userInitiated).async {
                dec.flush()
            }
        }
        assembler.reset()
        tcp.send("FPS:\(newFps)")
    }

    func switchScreen(_ index: Int) {
        guard state == .connected else { return }
        currentScreen = index

        // 取出旧解码器并置空（在锁内操作引用，锁外执行 stop 避免死锁）
        decoderLock.lock()
        let oldDecoder = decoder
        decoder = nil
        decoderLock.unlock()

        // stop() 内部会 Invalidate VT session，可能阻塞等待回调完成
        // 在后台线程执行防止主线程卡死
        if let old = oldDecoder {
            DispatchQueue.global(qos: .userInitiated).async {
                old.stop()
            }
        }

        assembler.reset()
        tcp.send("SCREEN:\(index)")
    }

    func sendEvent(_ event: ControlEvent) {
        guard state == .connected else { return }
        let str = event.serialize()
        let fid = nextFrameId()
        sender.sendEvent(str, frameId: fid)
    }

    // MARK: - Audio

    func setAudioEnabled(_ enabled: Bool) {
        guard enabled != audioEnabled else { return }
        audioEnabled = enabled
        if enabled, state == .connected {
            tcp.send("AUDIO_ON")
            audioCapture.start()
        } else {
            audioCapture.stop()
            if state == .connected { tcp.send("AUDIO_OFF") }
        }
    }

    // MARK: - File Transfer

    /// 发送文件列表到 Mac ~/Downloads
    func sendFiles(urls: [URL]) {
        guard state == .connected, !urls.isEmpty else { return }
        Task(priority: .userInitiated) {
            for url in urls {
                await sendSingleFile(url: url)
            }
            await MainActor.run { fileTransferProgress = nil }
        }
    }

    private func sendSingleFile(url: URL) async {
        let fileName = url.lastPathComponent
        guard let attrs = try? FileManager.default.attributesOfItem(atPath: url.path),
              let fileSize = attrs[.size] as? Int, fileSize > 0 else {
            NSLog("[FileTransfer] 无法获取文件大小: \(url.lastPathComponent)"); return
        }

        NSLog("[FileTransfer] 开始 \(fileName) \(fileSize) bytes")
        await MainActor.run { fileTransferProgress = 0 }

        tcp.send("FILE_START:\(fileName):\(fileSize)")

        // 等待 Mac 回复 FILE_READY（最多 5 秒）
        let deadline = Date().addingTimeInterval(5)
        while Date() < deadline {
            fileReadyLock.lock()
            let ready = fileReadySignal
            if ready { fileReadySignal = false }
            fileReadyLock.unlock()
            if ready { break }
            try? await Task.sleep(nanoseconds: 50_000_000)
        }

        // 流式发送文件字节
        guard let stream = InputStream(url: url) else { return }
        stream.open()
        defer { stream.close() }

        let chunkSize = 32_768  // 32KB
        var buf = [UInt8](repeating: 0, count: chunkSize)
        var sent = 0

        while stream.hasBytesAvailable {
            let read = stream.read(&buf, maxLength: chunkSize)
            guard read > 0 else { break }
            tcp.sendRaw(Data(buf[0..<read]))
            sent += read
            let progress = Double(sent) / Double(fileSize)
            await MainActor.run { fileTransferProgress = progress }
        }

        tcp.send("FILE_END")
        NSLog("[FileTransfer] 完成 \(fileName) \(sent) bytes")
    }

    // MARK: - Disconnect

    func disconnect() {
        guard state != .idle else { return }
        if tcp.isConnected { tcp.send("DISCONNECT") }
        Task { await teardown() }
    }

    // MARK: - TCP 命令处理

    private func handleTcpLine(_ line: String) {
        let t = line.trimmingCharacters(in: .whitespacesAndNewlines)

        if t.hasPrefix("RESOLUTION:") {
            let parts = t.dropFirst("RESOLUTION:".count).split(separator: ",")
            if let w = Int(parts.first ?? ""), let h = Int(parts.last ?? "") {
                remoteResolution = CGSize(width: w, height: h)
                videoView?.remoteSize = CGSize(width: w, height: h)
                // RESOLUTION 到达时无论之前是否收到 OK，都（重新）启动解码器
                // 这样能处理两种时序：
                //   · 初始连接：RESOLUTION → SCREENS → iOS发FPS → OK（解码器已在此启动）
                //   · 切屏：OK → startStreaming → RESOLUTION（先 OK 后 RESOLUTION，
                //     此时解码器为 nil，RESOLUTION 到达后才真正启动）
                startDecoderIfReady()
            }
        } else if t.hasPrefix("SCREENS:") {
            let list = t.dropFirst("SCREENS:".count).split(separator: ",").compactMap { entry -> ScreenInfo? in
                let parts = entry.split(separator: ":")
                guard parts.count >= 3,
                      let idx = Int(parts[0]) else { return nil }
                let name = String(parts[1])
                let dims = parts[2].split(separator: "x")
                let w = Int(dims.first ?? "0") ?? 0
                let h = Int(dims.last  ?? "0") ?? 0
                return ScreenInfo(id: idx, name: name, width: w, height: h)
            }
            if !list.isEmpty { screens = list }
        } else if t == "OK" {
            // Mac 在两个场景发 OK：
            //   1. FPS 命令回应（初始连接）：之后会收到 RESOLUTION，届时启动解码器
            //   2. SCREEN 切换回应：紧接着 startStreaming，再发 RESOLUTION
            // 两种情况都在 RESOLUTION 处理里统一启动解码器，这里不提前启动，
            // 避免使用旧分辨率创建解码器后被 RESOLUTION 覆盖。
            NSLog("[Session] 收到 OK，等待 RESOLUTION 后启动解码器")
        } else if t == "PONG" {
            // 心跳回应，忽略
        } else if t == "FILE_READY" {
            fileReadyLock.lock()
            fileReadySignal = true
            fileReadyLock.unlock()
        } else if t.hasPrefix("FILE_OK:") {
            let name = String(t.dropFirst("FILE_OK:".count))
            NSLog("[FileTransfer] Mac 已接收: \(name)")
        } else if t.hasPrefix("FILE_ERR:") {
            NSLog("[FileTransfer] Mac 报错: \(t.dropFirst("FILE_ERR:".count))")
        }
    }

    // MARK: - Decoder

    private func startDecoderIfReady() {
        decoderLock.lock()
        defer { decoderLock.unlock() }
        // 若 videoView 尚未注入（onViewReady 还未回调），延迟到主线程下一 runloop 再试一次
        guard let v = videoView else {
            DispatchQueue.main.async { [weak self] in
                guard let self else { return }
                self.decoderLock.lock()
                defer { self.decoderLock.unlock() }
                guard let v = self.videoView else { return }
                self.startDecoderLocked(view: v)
            }
            return
        }
        startDecoderLocked(view: v)
    }

    private func startDecoderLocked(view: VideoSurfaceView) {
        guard decoder == nil else { return }
        let size = remoteResolution ?? CGSize(width: 1920, height: 1080)
        let dec  = H265Decoder(fps: fps)
        dec.displayLayer = view.displayLayer   // 直接绑定到 AVSampleBufferDisplayLayer
        dec.start()
        decoder = dec
        NSLog("[Session] 解码器启动 \(Int(size.width))x\(Int(size.height)) @ \(fps)fps")
    }

    // MARK: - Heartbeat

    private func startHeartbeat() {
        heartbeatTask?.cancel()
        heartbeatTask = Task {
            while !Task.isCancelled, state == .connected {
                try? await Task.sleep(nanoseconds: 30_000_000_000)  // 30s
                guard state == .connected else { break }
                tcp.send("PING")
            }
        }
    }

    // MARK: - Teardown

    @MainActor
    private func teardown() {
        heartbeatTask?.cancel(); heartbeatTask = nil

        if audioEnabled { tcp.send("AUDIO_OFF") }
        audioCapture.stop()
        audioEnabled = false
        macAudioPlayer.stop()

        tcp.disconnect()
        receiver.stop()

        decoderLock.lock()
        let oldDecoder = decoder
        decoder = nil
        decoderLock.unlock()
        // stop 在后台避免主线程死锁
        if let old = oldDecoder {
            DispatchQueue.global(qos: .userInitiated).async { old.stop() }
        }

        videoView?.flush()
        assembler.reset()
        fileReadyLock.lock()
        fileReadySignal = false
        fileReadyLock.unlock()
        fileTransferProgress = nil

        remoteResolution = nil
        screens          = []
        currentScreen    = 0
        device           = nil

        state = .disconnected

        endBackgroundTask()
    }

    // MARK: - Frame ID

    private func nextFrameId() -> UInt32 {
        eventFrameId = (eventFrameId &+ 1) & 0x7FFFFFFF
        return eventFrameId
    }

    // MARK: - Background Task

    private func beginBackgroundTask() {
        bgTask = UIApplication.shared.beginBackgroundTask(withName: "LowRemote-Session") { [weak self] in
            self?.endBackgroundTask()
        }
    }

    private func endBackgroundTask() {
        guard bgTask != .invalid else { return }
        UIApplication.shared.endBackgroundTask(bgTask)
        bgTask = .invalid
    }
}
